### PR TITLE
[stdlib] Set, Dictionary: Make Cocoa indices resilient for now

### PR DESCRIPTION
**Pull request by lorentey**
This makes `_CocoaSet.Index` and `_CocoaDictionary.Index` fully resilient for now, allowing us to replace/fix its internals at any point in the future.

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
cc @airspeedswift What do you think? This leaves the door open for a Cocoa indexing overhaul.

**Comment by lorentey** [comment]
Gosh; `Dictionary.Keys` and `.Values` uses `IndexingIterator`. That won't do at all.

**Comment by lorentey** [comment]
I think #19576 (for swapAt) and #19602 (for Histogram et al) will cut down these performance regressions.

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
Hm, I'll have to make the Cocoa Index storage representation `@_fixed_layout` to reduce these regressions.

I'll do that by switching it to a CoW type, which I was planning to do anyway.

**Comment by lorentey** [comment]
@swift-ci please benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
Oops, I forgot to apply these changes to Dictionary; I'm afraid these next benchmarks will still disappoint

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
Sigh

@swift-ci please smoke test linux platform

**Comment by lorentey** [comment]
Okay, this should do the trick. (He said, with a naive look on his face.)

File: stdlib/public/core/Dictionary.swift [hunk 1]
```diff
@@ -1897,7 +1897,7 @@ extension Dictionary.Index: Comparable {
 }
 
 extension Dictionary.Index: Hashable {
-  @inlinable
+  @_effects(readonly) // FIXME(cocoa-index): Make inlinable
```
Comment (jrose-apple) [inline stdlib/public/core/Dictionary.swift]: Is this correct? It writes to the hasher, after all.

**Comment by lorentey** [comment]
Oh hi! I'm just sitting over here, improving benchmarks by making things non-inlinable.

(This probably just means we lack indexing benchmarks for bridged sets/dictionaries.)

File: stdlib/public/core/Dictionary.swift (hunk 1 above)
Comment (lorentey) [reply stdlib/public/core/Dictionary.swift]: Oh, oops, good catch!

**Pull request merged**